src/rust/catloop/socket.rs

```diff
@@ -163,73 +146,57 @@ impl SharedMemorySocket {
+                pop_request_id(catmem.clone(), self.catmem_qd.expect("should be connected")).await?;
+
+            // Received a request id so create the new connection. This involves create the new duplex pipe
+            // and sending the port number to the remote.
+            if self.pending_request_ids.contains(&request_id) {
+                debug!("do_accept(): duplicate request (request_id={:?})", request_id.0);
+                continue;
+            } else {
+                self.pending_request_ids.insert(request_id);
+                break create_pipe(
+                    self.catmem_qd.expect("pipe should have been created"),
+                    catmem.clone(),
+                    &ipv4,
+                    new_port,
+                )
+                .await?;
+            }
+        };
 
-            let new_addr: SocketAddrV4 = SocketAddrV4::new(ipv4, new_port);
-            let new_socket: Self = Self::alloc(new_qd, None, Some(new_addr));
-            let result: Result<RequestId, Fail> = self.pop_request_id(&mut runtime, &mut catmem, new_qd).await;
-
-            // Check that the remote has retrieved the port number and responded with a valid request id.
-            match result {
-                // Valid response. Connection successfully established, so return new port and pipe to application.
-                Ok(request_id) => {
-                    // If we've never seen this before, something has gone very wrong.
-                    assert!(self.pending_request_ids.contains(&request_id));
-                    self.pending_request_ids.remove(&request_id);
-                    return Ok((new_socket, new_addr.into()));
-                },
-                // Some error.
-                Err(e) => {
-                    // Clean up newly allocated duplex pipe.
-                    catmem.close(new_qd)?;
-                    runtime.free_ephemeral_port(new_port)?;
-                    return Err(e);
-                },
-            };
-        }
+        let new_addr: SocketAddrV4 = SocketAddrV4::new(ipv4, new_port);
+        let new_socket: Self = Self::alloc(new_qd, None, Some(new_addr));
+
+        // Check that the remote has retrieved the port number and responded with a valid request id.
+        match pop_request_id(catmem.clone(), new_qd).await {
+            // Valid response. Connection successfully established, so return new port and pipe to application.
+            Ok(request_id) => {
+                // If we've never seen this before, something has gone very wrong.
+                assert!(self.pending_request_ids.contains(&request_id));
+                self.pending_request_ids.remove(&request_id);
+                return Ok((new_socket, new_addr.into()));
+            },
+            // Some error.
+            Err(e) => {
+                // Clean up newly allocated duplex pipe.
+                catmem.close(new_qd)?;
+                return Err(e);
+            },
+        };
```

Comment:
There is no need for return statements here.

Reply:
yup, removed